### ISSUE TITLE
:coda_read_definitions_fuzzer: Out-of-memory in coda_read_definitions_fuzzer

### ISSUE THREAD

**Issue by schwehr** [stcorp/coda#73]
```
==3601590== ERROR: libFuzzer: out-of-memory (malloc(4278537040))
  | #8 0x55a3a29a93b3 in parse_entry third_party/stcorp_coda/libcoda/coda-definition-parse.c:4152:19
  | #9 0x55a3a29aa619 in read_definition_file third_party/stcorp_coda/libcoda/coda-definition-parse.c:4253:9
  | #10 0x55a3a29aa401 in coda_read_definitions third_party/stcorp_coda/libcoda/coda-definition-parse.c:4461:21
  | #11 0x55a3a2962b49 in LLVMFuzzerTestOneInput third_party/stcorp_coda/fuzz/coda_read_definitions_fuzzer.cc:19:3
```

```c
    parser_info_init(&info);
    info.zf = zf;
    info.entry_base_name = name;
    info.product_class = current_product_class;
    info.product_definition = current_product_definition;

    filesize = za_get_entry_size(entry);
    info.buffer = malloc(filesize);   // <-- No sanity check on filesize before use
    if (info.buffer == NULL)
```

[testcase-5666406165381120.zip](https://github.com/stcorp/coda/files/4644434/testcase-5666406165381120.zip)

**Comment by svniemeijer** [comment]
How is this an issue?
In general every malloc could result in an out-of-memory.
The general approach is to try and nicely bail out if this fails, which is what we are doing (if bail out + cleanup is not valid, then that would be an actual issue).

**Comment by schwehr** [comment]
A bad malloc size doesn't always return a null in many environments.  Many of these are denial of service type attacks.  A concrete example:

1. A cloud instance might have requested 4GB of RAM.  It's processing small files, but a corrupt zip of less than 1K causes it to try to allocate just less than the available ram or the process runner just increase the chance of a worker being killed that uses more ram than requested.  Then the worker faces an increase chance of being killed by the cluster manager.  If this happens a lot, it can starve a service.
1. Given a machine with a large amount of ram and swap, say 200GB RAM and a couple TB swap meant for processing really large assets.  If this machine gets a malloc request of a TB, that might be okay for the large asset, but if the definition file does that, it's a problem. 
1. For fuzzers, they are definitely see this as a problem.  It slows down fuzzing for un-interesting cases.  That goes away once there is a reasonable limit.  e.g. if the fuzzer sees > 2GB of RAM usage for an input less than 4KB, it will complain every time.

Overly large allocs can cause undo resource usage.  That can stop a process from making progress, unexpectedly use up quota, and/or cause larger than expected billing in a cloud situation.
 
Given a zip of size X there is a reasonable max for the compression ratio of a valid file.  A common solution is to let the caller specify an optional max amount of ram to allocate.  e.g. with an environment variable.  Another example is the Kakadu JPEG2000 library went for a much more complicated memory pools solution, but I don't think that is worth it for coda and harp.

The source of these allocs can be corrupted files by accident (bad bits and/or truncated files), buggy software writing bad files, or malicious actors.

Most users probably won't care, but I definitely do and others might.  An error message saying that there are not enough resources available is much easier to debug than the alternative.

**Comment by svniemeijer** [comment]
I don't find these arguments that convincing.

Note that we _are_ providing an error message saying that there are not enough resources available (which is when and exactly when a malloc no longer succeeds).

What you are asking for is that a library performs some form of QoS on its allocs, considering some more important than others and introducing arbitrary limits. This is not something that we will add to our code.

What I would still consider is to introduce checks on the file content where it can be determined that the value is really _invalid_. And then introduce this check before we get to the situation of performing a malloc.
This will give you the same result, but it is conceptually a very different discussion. The discussion should then be about determining the proper validity range of a value that was read. It should not be about putting arbitrary limits on mallocs.

**Comment by schwehr** [comment]
I will patch our local copy of coda.  A bummer to have our local copy diverge from master.

**Comment by schwehr** [comment]
For anyone else who might be interested.  filesize is 4278537040 with testcase-5666406165381120.  And you need to unzip testcase-5666406165381120.zip first.  testcase-5666406165381120 is a zip

```c
    if (filesize > 10000000) {
      coda_set_error(CODA_ERROR_OUT_OF_MEMORY,
                     "filesize unreasonably large (%lu bytes) (%s:%d)",
                     filesize, __FILE__, __LINE__);
        parser_info_delete(&info);
        return -1;
    }
```

**Comment by svniemeijer** [comment]
I don't know why you closed this issue.

The fact that your solution is not the correct one, does not mean that we can't introduce a proper solution for this.
There are proper upper limits that we can check against for both the compressed size (should be smaller than the file size) and the uncompressed size (can be compared against the compressed size using the theoretical upper compression factor limit of 1032).

However, if I am running your test file I am getting:
```
% codadd -D testcase-5666406165381120
ERROR: could not read data from definition file: error in zip file (local header has incorrect signature)
```
So I don't understand how you get to the stage of this excessive malloc.

**Comment by svniemeijer** [comment]
Never mind. I can see the large malloc (it just succeeds).

**Comment by schwehr** [comment]
I closed it because it seems like you were talking about a different issue.  It's fine to switch topics and if you have a patch for the max theoretical size, then great.  A link to a reference that explains the max would be great.

However, I see at least 2 separate issues.  I was looking for a max expected filesize in a zip for coda/harp.  I picked 10 million bytes.

Try running the test with various limits.  The value in the zip is 4278537040   e.g. a mem limit of 1GB up to ~4GB should get a nullptr back from the malloc.  A values just above that value might return a block of memory, but later run out of memory.  Another thing to try is setup a docker instance with very constrained ram (e.g. 1GB ) and 100's of GB of swap.  Anything that can malloc 4+ GB of swap in one go is going to likely make the process crazy slow effectively doing a DOS.

Which begs the question of what are the expected roles for zip in coda?  Is it just definitions or can one or more grib or zip files be embedded inside a zip?  This is where some basic CI tests would make it obvious to the reader what is the expect functionality of the zip code here.  Test cases are a great way to state expectations of the codes usage and capabilities.
